src/thread.rs

```diff
@@ -0,0 +1,596 @@
+            trace!(
+                "{:?} blocked on {:?} when trying to join",
+                self.active_thread,
+                joined_thread_id
+            );
+        }
+        Ok(())
+    }
+
+    /// Set the name of the active thread.
+    fn set_thread_name(&mut self, new_thread_name: Vec<u8>) {
+        self.active_thread_mut().thread_name = Some(new_thread_name);
+    }
+
+    /// Get the name of the active thread.
+    fn get_thread_name(&mut self) -> InterpResult<'tcx, Vec<u8>> {
+        if let Some(ref thread_name) = self.active_thread_mut().thread_name {
+            Ok(thread_name.clone())
+        } else {
+            throw_ub_format!("thread {:?} has no name set", self.active_thread)
+        }
+    }
+
+    /// Allocate a new blockset id.
+    fn create_blockset(&mut self) -> BlockSetId {
+        self.blockset_counter = self.blockset_counter.checked_add(1).unwrap();
+        self.blockset_counter.into()
+    }
+
+    /// Block the currently active thread and put it into the given blockset.
+    fn block_active_thread(&mut self, set: BlockSetId) {
+        let state = &mut self.active_thread_mut().state;
+        assert_eq!(*state, ThreadState::Enabled);
+        *state = ThreadState::Blocked(set);
+    }
+
+    /// Unblock any one thread from the given blockset if it contains at least
+    /// one. Return the id of the unblocked thread.
+    fn unblock_some_thread(&mut self, set: BlockSetId) -> Option<ThreadId> {
+        for (id, thread) in self.threads.iter_enumerated_mut() {
+            if thread.state == ThreadState::Blocked(set) {
+                trace!("unblocking {:?} in blockset {:?}", id, set);
+                thread.state = ThreadState::Enabled;
+                return Some(id);
+            }
+        }
+        None
+    }
+
+    /// Change the active thread to some enabled thread.
+    fn yield_active_thread(&mut self) {
+        self.yield_active_thread = true;
+    }
+
+    /// Decide which action to take next and on which thread.
+    ///
+    /// The currently implemented scheduling policy is the one that is commonly
+    /// used in stateless model checkers such as Loom: run the active thread as
+    /// long as we can and switch only when we have to (the active thread was
+    /// blocked, terminated, or was explicitly asked to be preempted).
+    fn schedule(&mut self) -> InterpResult<'tcx, SchedulingAction> {
+        if self.threads[self.active_thread].check_terminated() {
+            // Check if we need to unblock any threads.
+            for (i, thread) in self.threads.iter_enumerated_mut() {
+                if thread.state == ThreadState::BlockedOnJoin(self.active_thread) {
+                    trace!("unblocking {:?} because {:?} terminated", i, self.active_thread);
+                    thread.state = ThreadState::Enabled;
+                }
+            }
+            return Ok(SchedulingAction::ExecuteDtors);
```

Comment:
Interestingly, this does *not* change to a different active thread. Seems like there is a risk of an endless loop here if something else does not switch away? Where is that switching happening?

Reply:
Endless loop could happen only if destructors are scheduled infinitely. Otherwise, after the last destructor is executed, the thread remains in `Terminated` state and `check_terminated` returns `false`, thus the thread does not get scheduled anymore.

Reply:
Ah, I forgot that `check_terminated` returns `false` when the thread is already terminated. That is not clear at all from the name (it sounds like it would just test if `state == Terminated`).

Please at least add a comment along those lines here above the call to `check_terminated`. Ideally we'd find a better name for that function, but I cannot think of one now.